File: apps/web/src/lib/rut.js
/**
 * RUT chileno: validación por módulo 11 y formato mientras se escribe.
 *
 * Esto es para la experiencia de quien completa el formulario —avisar del dedazo antes de
 * mandar, y poner los puntos solo—, no para la seguridad. La validación que manda vive en
 * la API (`app/core/rut.py`), porque el checkout es una llamada HTTP como cualquier otra
 * y nada impide mandarla a mano.
 *
 * Lo que ninguna de las dos hace: confirmar que el RUT EXISTA en el SII o que sea de la
 * empresa que dice el formulario. Valida la forma, no la identidad.
 */

const soloValidos = (valor) => String(valor || '').replace(/[^0-9kK]/g, '').toUpperCase();

const digitoVerificador = (cuerpo) => {
  let suma = 0;
  let factor = 2;
  for (let i = cuerpo.length - 1; i >= 0; i -= 1) {
    suma += Number(cuerpo[i]) * factor;
    // Los factores van 2,3,4,5,6,7 y vuelven a empezar.
    factor = factor === 7 ? 2 : factor + 1;
  }
  const resto = 11 - (suma % 11);
  if (resto === 11) return '0';
  if (resto === 10) return 'K';
  return String(resto);
};

/** ¿Está bien formado y su dígito verificador cuadra? */
export const rutEsValido = (valor) => {
  const limpio = soloValidos(valor);
  // Menos de 7 dígitos de cuerpo no es un RUT chileno vigente; más de 8 tampoco.
  if (limpio.length < 8 || limpio.length > 9) return false;

  const cuerpo = limpio.slice(0, -1);
  const verificador = limpio.slice(-1);
  if (!/^\d+$/.test(cuerpo)) return false;

  return digitoVerificador(cuerpo) === verificador;
};

/**
 * Con puntos y guion, para mostrar mientras se escribe: "12.345.678-5".
 *
 * Se aplica en cada tecla, así que tiene que aguantar un RUT a medio escribir sin
 * estorbar: con menos de dos caracteres devuelve lo que haya, tal cual.
 */
export const formatearRut = (valor) => {
  const limpio = soloValidos(valor);
  if (limpio.length < 2) return limpio;

  const cuerpo = limpio.slice(0, -1);
  const verificador = limpio.slice(-1);
  const conPuntos = cuerpo.replace(/\B(?=(\d{3})+(?!\d))/g, '.');

  return `${conPuntos}-${verificador}`;
};

/** Sin puntos y con guion, que es como lo guarda la API. */
export const normalizarRut = (valor) => {
  const limpio = soloValidos(valor);
  if (limpio.length < 2) return limpio;
  return `${limpio.slice(0, -1)}-${limpio.slice(-1)}`;
};

File: apps/api/app/core/rut.py
"""
Validacion y normalizacion de RUT chileno.

El digito verificador se calcula por modulo 11. Sirve para atajar el error de tipeo, que
es el caso real: alguien escribe su RUT apurado en el celular, se equivoca en un digito y
la factura sale a nombre de otra empresa o directamente no se puede emitir.

Lo que NO hace, y conviene tener claro: un RUT puede pasar el modulo 11 y no existir en el
SII, o existir y no corresponder a la empresa que dice el formulario. Esto valida la
FORMA, no la identidad.
"""
import re

_LIMPIEZA = re.compile(r"[^0-9kK]")


def _digito_verificador(cuerpo: str) -> str:
    """Calcula el digito que le corresponde a un cuerpo de RUT, por modulo 11."""
    suma = 0
    factor = 2
    for digito in reversed(cuerpo):
        suma += int(digito) * factor
        # Los factores van 2,3,4,5,6,7 y vuelven a empezar.
        factor = 2 if factor == 7 else factor + 1

    resto = 11 - (suma % 11)
    if resto == 11:
        return "0"
    if resto == 10:
        return "K"
    return str(resto)


def normalizar_rut(valor: str) -> str:
    """
    Deja el RUT en la forma en que se guarda: sin puntos, en mayuscula y con guion.

    "76.123.456-7", "761234567" y "76123456-K" entran igual y salen como "76123456-7" /
    "76123456-K". Guardar siempre igual evita que el mismo RUT quede escrito de tres
    formas distintas segun como lo tipeo cada cliente.
    """
    limpio = _LIMPIEZA.sub("", valor or "").upper()
    if len(limpio) < 2:
        return limpio
    return f"{limpio[:-1]}-{limpio[-1]}"


def rut_es_valido(valor: str) -> bool:
    """True si el RUT esta bien formado y su digito verificador cuadra."""
    limpio = _LIMPIEZA.sub("", valor or "").upper()

    # Menos de 7 digitos de cuerpo no es un RUT chileno vigente; mas de 8 tampoco.
    if not 8 <= len(limpio) <= 9:
        return False

    cuerpo, verificador = limpio[:-1], limpio[-1]
    if not cuerpo.isdigit():
        return False

    return _digito_verificador(cuerpo) == verificador


def formatear_rut(valor: str) -> str:
    """Con puntos y guion, para mostrar: "76.123.456-7". Solo para lectura."""
    limpio = _LIMPIEZA.sub("", valor or "").upper()
    if len(limpio) < 2:
        return valor or ""

    cuerpo, verificador = limpio[:-1], limpio[-1]
    con_puntos = ""
    for i, digito in enumerate(reversed(cuerpo)):
        if i and i % 3 == 0:
            con_puntos = "." + con_puntos
        con_puntos = digito + con_puntos

    return f"{con_puntos}-{verificador}"
